### PR TITLE
[Bugfix] _add_filters_from_pre_query doesn't handle dim specs

### PR DESCRIPTION
**Pull request by Mogball**
Ok this should address the underlying issue behind https://github.com/apache/incubator-superset/issues/3943. `_add_filter_from_pre_query_data` wasn't at all handling dimension specs. Also added some better error checking on modifying dimension specs.

edit: https://github.com/apache/incubator-superset/issues/3889 similar issue

### CODE REVIEW

**Comment by ghulands** [comment]
Thanks @Mogball for taking a look at this. I just applied the patch and tested it out. I now get "No data was returned" instead of the exception. If I click on View Query it only shows the first phase of the query and not the second. When I run that query directly against druid, I get results for the topN.

**Comment by Mogball** [comment]
@ghulands Thanks for letting me know. I removed a condition thinking Druid handles single `"dimension"` specs but I think I was right before. Checkout the change and let me know if it's working.

**Comment by Mogball** [comment]
Could you post logs from running these queries? (And it's still just doing 1 phase right?)

**Comment by ghulands** [comment]
Yes, it's still only running 1 phase. I edited the log to reduce the map size of the dimension.

```
2017-12-01 19:07:43,675:DEBUG:root:[36m[stats_logger] (incr) explore_json[0m
2017-12-01 19:07:43,764:DEBUG:root:[36m[stats_logger] (incr) loaded_from_source[0m
2017-12-01 19:07:43,784:INFO:root:Running groupby query for dimensions [[{u'outputName': u'prettyName', u'extractionFn': {u'outputName': u'dimensionOutputName', u'lookup': {u'map': {u'Code1': u'PrettyName1', u'Code2': u'PrettyName2'}, u'type': u'map', u'isOneToOne': False}, u'retainMissingValue': False, u'type': u'lookup', u'replaceMissingValueWith': u'missing_value', u'dimension': u'dimensionName'}, u'type': u'extraction', u'dimension': u'codeName', u'outputType': u'STRING'}]]
2017-12-01T19:07:43.784623+00:00 orchardapp710899[app_siri-ops_siri-superset-prod.web.v34]: 2017-12-01 19:07:43,784:INFO:root:Running two-phase query for timeseries
```

**Comment by Mogball** [comment]
@ghulands 

Okay I think I figured it out. Extraction functions that change the name of dimension values (in your case, mapping from a code to a name) cause issues with the two-phase query because it builds additional filters from output values (which are different from the original values). `_add_filters_from_pre_query` will ignore extraction functions. Give it another go.

**Comment by ghulands** [comment]
Hi @Mogball, This now works! Thanks so much for the quick iterating on it.

**Comment by Mogball** [comment]
@fabianmenges RFC on this solution

For two-phase queries, would it be better to remove extraction functions from pre-queries (and just use the dimension name) and add them back for the final query?

**Comment by fabianmenges** [comment]
I think it makes sense to strip the pre-queries (phase 1 queries) from the extraction functions since they will only change how data is dispalyed and pre-queries won't be displayed.

**Comment by Mogball** [comment]
Should we replace any elements in `qry['dimensions']` that are dimension specs with `dim_spec['dimension']` (on which they operate) for pre-queries?

File: superset/connectors/druid/views.py [hunk 1]
```diff
@@ -61,9 +62,26 @@ class DruidColumnInlineView(CompactCRUDMixin, SupersetModelView):  # noqa
             True),
     }
 
+    def pre_update(self, col):
+        # If a dimension spec JSON is given, ensure that it is
+        # valid JSON and that `outputName` is specified
+        if col.dimension_spec_json:
```
Comment (fabianmenges) [inline superset/connectors/druid/views.py]: While I like the validation here in general, I feel like its a bit arbitrary that we validate that 
`dimension_spec_json` is valid but we don't do any validation for SQL columns or SQL/Druid metrics.

Comment (Mogball) [reply superset/connectors/druid/views.py]: I noticed as well. Should we starting adding validation (or 'warnings') for other things as well?

File: superset/connectors/druid/models.py [hunk 2]
```diff
@@ -1039,7 +1061,18 @@ def run_query(  # noqa / druid
...
+                pre_qry_dims = []
+                # Replace dimensions specs with their `dimension`
+                # values, and ignore those without
+                for dim in qry['dimensions']:
+                    if isinstance(dim, dict):
+                        if 'dimension' in dim:
+                            pre_qry_dims.append(dim['dimension'])
+                    else:
+                        pre_qry_dims.append(dim)
+                pre_qry['dimensions'] = utils.unique_list(pre_qry_dims)
```
Comment (fabianmenges) [inline superset/connectors/druid/models.py]: Any reason your not just doing this `list(set(pre_qry_dims))`

File: superset/utils.py [hunk 3]
```diff
@@ -746,3 +746,9 @@ def get_filter_key(f):
                     form_data['filters'] += [filtr]
         # Remove extra filters from the form data since no longer needed
         del form_data['extra_filters']
+
+def unique_list(target_list):
```
Comment (fabianmenges) [inline superset/utils.py]: Don't think this helper is useful

Comment (mistercrunch) [reply superset/utils.py]: the easy way to do this is `list(set(list))` though you can usually just use a set from the get go and stick to it

Comment (Mogball) [reply superset/utils.py]: true didn't realise it was that simple 🥇

**Review by fabianmenges — APPROVED** [review]

**Comment by chirpy2291** [comment]
Hi, i am facing the same issue.On running single custom dimension i get the following error:
KeyError: u"None of [[u'outputName', u'extractionFn', u'type', u'outputType', u'dimension']] are in the [index]"

Can you specify which version of druid to use? and what is the fix? @Mogball does your fix solve this?

**Comment by Mogball** [comment]
What version of superset are you using? I think this is in version 0.21.0

**Comment by chirpy2291** [comment]
Does 0.21 solve both the mentioned problems?


I am using 0.20.
Extensions are non filterable.And single extesion dimension is non
groupable.Any reason why topN query doesnt work on superset for single
extension dimension?
Same works if i hit the same query on druid.

**Comment by Mogball** [comment]
I'm pretty sure those issues we're ironed out by then. Give 0.21.0 a try

**Comment by chirpy2291** [comment]
Hi @Mogball that was great help.I am now able to group the custom dimensions as well as view them on charts.

Do we have any update/workaround for filtering the same custom extension.I am able to only filter them through Result filters only.

**Comment by Mogball** [comment]
Not sure about that. Make an issue if there isn't one already.

**Pull request merged**